Sources/Document/Json/JSONArray.swift

```diff
@@ -16,16 +16,545 @@
+        return toWrappedElement(from: inserted)
+    }
+
+    /**
+     * `insertBefore` inserts a value before the given next element.
+     */
+    func insertBefore(nextID: TimeTicket, value: Any) throws -> Any? {
+        let inserted = try insertBeforeInternal(nextCreatedAt: nextID, value: value)
+        return toWrappedElement(from: inserted)
+    }
+
+    /**
+     * `moveBefore` moves the element before the given next element.
+     */
+    func moveBefore(nextID: TimeTicket, id: TimeTicket) throws {
+        try self.moveBeforeInternal(nextCreatedAt: nextID, createdAt: id)
+    }
+
+    /**
+     * `moveAfter` moves the element after the given previous element.
+     */
+    func moveAfter(previousID: TimeTicket, id: TimeTicket) throws {
+        try self.moveAfterInternal(previousCreatedAt: previousID, createdAt: id)
+    }
+
+    /**
+     * `moveFront` moves the element before the first element.
+     */
+    func moveFront(id: TimeTicket) throws {
+        try self.moveFrontInternal(createdAt: id)
+    }
+
+    /**
+     * `moveLast` moves the element after the last element.
+     */
+    func moveLast(id: TimeTicket) throws {
+        try self.moveLastInternal(createdAt: id)
+    }
+
+    subscript(index: Int) -> Any? {
+        guard let value = try? self.target.get(index: index) else {
+            return nil
+        }
+        return toJSONElement(from: value)
+    }
+
+    @discardableResult
+    func append(_ value: Any) -> Int {
+        self.push(value)
+    }
+
+    func append(values: [Any]) {
+        self.push(values: values)
+    }
+
+    func push(values: [Any]) {
+        values.forEach { value in
+            if let value = value as? [String: Any] {
+                let appendedIndex = push(JSONObject())
+                let jsonObject = self[appendedIndex] as? JSONObject
+                jsonObject?.set(value)
+            } else if let value = value as? [Any] {
+                let appendedIndex = push(JSONArray())
+                let jsonArray = self[appendedIndex] as? JSONArray
+                jsonArray?.push(value)
+            } else {
+                push(value)
+            }
+        }
+    }
+
+    @discardableResult
+    func push(_ value: Any) -> Int {
+        self.pushInternal(value)
+    }
+
+    func length() -> Int {
+        return self.target.length
+    }
+
+    func remove(index: Int) -> CRDTElement? {
+        Logger.trivial("array[\(index)]")
+        return self.removeInternal(byIndex: index)
+    }
+
+    /**
+     * `pushInternal` pushes the value to the target array.
+     */
+    @discardableResult
+    private func pushInternal(_ value: Any) -> Int {
+        let appendedElement = try? self.insertAfterInternal(previousCreatedAt: self.target.getLastCreatedAt(), value: value)
+        guard appendedElement != nil else {
+            return Self.notAppend
+        }
+        return self.target.length - 1
```

Comment:
```self.target.length``` can not be zero based index, so I added ```-1```, but ```ArrayProxy.pushInternal``` returns target.length. What is the correct answer?

Reply:
We follow the interface of JS.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push#return_value